### ISSUE TITLE
MissingTemplateParam triggered by calling code

### ISSUE THREAD

**Issue by neclimdul** [vimeo/psalm#6870]
Psalm version 4.12.0

ERROR: MissingTemplateParam - app/Services/ReportBuilder.php:126:19 - ArrayObject has missing template params, expecting 2 (see https://psalm.dev/182)
      $drawings = $sheeet->getDrawingCollection();

The problem is with the PhpSpreadsheet [Worksheet](https://github.com/PHPOffice/PhpSpreadsheet/blob/master/src/PhpSpreadsheet/Worksheet/Worksheet.php#L542) class which only templates one argument.

I tried to make a simplified test case for this and it didn't work. This was the attempt.
https://psalm.dev/r/1a3c9bae46

**Comment by orklah** [comment]
This is a legit error, in fact, PHPStan outputs the same on the PhpSpreadsheet repo: https://github.com/PHPOffice/PhpSpreadsheet/blob/master/phpstan-baseline.neon#L6379 but it was ignored.

Best way to fix this: create a PR to their repo or create a stub for this that will override the phpdoc.

(On a side note: I'm so happy to see PHPSpreasheet embracing static analysis! Kudos @PowerKiKi!)

**Comment by neclimdul** [comment]
Yeah no doubt an error and an easy fix. Should calling code trigger errors in code the caller can't control though?

**Comment by orklah** [comment]
It directly impacts your code though. What if you call this method and it returns a corrupted type?

**Comment by neclimdul** [comment]
A fair point. Holding back the rest of your code bases quality because you're fighting the documentation on an external library you might not be able to affect leaves a bad taste in my mouth though.

Assuming nothing changed though, the error is not very helpful. There is no reference to ArrayObject or variable templates anywhere near the calling code. So instead of a 5 second suppression I had to go on a spelunking journey through rule documentation eventually stumbling into the fact that the return type documentation in the library was updated to an ArrayObject.

**Comment by orklah** [comment]
The error is emitted here: https://github.com/vimeo/psalm/blob/b29737f2b154748430250effdbcf47de8a5904df/src/Psalm/Internal/TypeVisitor/TypeChecker.php#L225 in a method that is shared with every place where we have templates to check (in fact it's not even a template specific piece of code, it's the function called when we have a named object to check (it may be `stdClass`, `ArrayObject<int>` or `MySuperClass`). In the current code, we don't know if we're on a return type, a param type or an inline @var somewhere in the code. We also don't know if the named object comes from vendors or not.

While keeping this in mind, is there a change in the error message that would have helped you?

**Comment by neclimdul** [comment]
I'm not sure how to answer. Obviously it would be useful to reference the line that actually has the error, in this case the Worksheet.php file. Sounds like you're saying all the context has been thrown away? I tried stepping through but there are too many properties to grok and couldn't locate where that would be stored.

**Comment by orklah** [comment]
Oh, nvm, I thought we were in a static class.

We actually have access to `$this->code_location` who contains a CodeLocation object which knows the file_name.

You could use that to include the file name on the error if you want. Up for a PR?

**Comment by neclimdul** [comment]
Sure, I'l take a look.

**Comment by PowerKiKi** [comment]
Try this: https://github.com/PHPOffice/PhpSpreadsheet/pull/2384

**Comment by neclimdul** [comment]
I got stalled on this because I couldn't find any coverage and tweaks to the string didn't cause any failures. 

Do you have a pointer on where such a test should live and what they would look like? Want to make sure I don't break anything but also that tests match the expect format and style.

**Comment by SamMousa** [comment]
https://psalm.dev/r/ce83fdc437

I ran into the same issue, and have found a minimal way to reproduce.

The error message seems to just be 100% wrong as the premise false:

> Emitted when using the `@extends/@implements` annotation to extend a class without extending all its template params.

I've used neither of those annotations.
